File: examples/other_releases/4.4.3.0/navigate/ios/Venues/Venues/DrawingSwitcher.swift
/*
* Copyright (C) 2020 HERE Europe B.V.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*
* SPDX-License-Identifier: Apache-2.0
* License-Filename: LICENSE
*/

import heresdk
import UIKit

// Allows to select a drawing inside a venue trough UI.
public class DrawingSwitcher: UIView {
    @IBOutlet private var contentView: UIView!
    @IBOutlet private weak var selectedDrawingButton: UIButton!
    @IBOutlet private weak var drawingsTableView: UITableView!

    private var venueMap: VenueMap?

    private static let minHeight = CGFloat(30)
    static let maxHeight = CGFloat(140)

    private weak var venueMapDelegate: DrawingSwitcherDelegate?
    private var heightConstraint: NSLayoutConstraint?
    private var drawingNames: [String] = []

    init(_ frame: CGRect) {
        super.init(frame: frame)
        customInit()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        customInit()
    }

    deinit {
        removeVenueMapDelegate()
    }

    func customInit() {
        let frameworkBundle: Bundle = Bundle(for: DrawingSwitcher.self)
        frameworkBundle.loadNibNamed("DrawingSwitcher", owner: self, options: nil)
        addSubview(contentView)
        // Set up a visual style of DrawingSwitcher
        contentView.frame = bounds
        contentView.layer.borderWidth = 1.0
        contentView.layer.borderColor = UIColor.gray.cgColor
        updateViewForSelectedVenue()

        // Set up a table view with drawing names.
        drawingNames = [String]()
        drawingsTableView.dataSource = self
        drawingsTableView.delegate = self
        drawingsTableView.register(UITableViewCell.self, forCellReuseIdentifier: "DrawingSwitcherCellID")

        // Set up a visual style of title, which contains information about the selected drawing.
        selectedDrawingButton.titleLabel?.numberOfLines = 1
        selectedDrawingButton.titleLabel?.adjustsFontSizeToFitWidth = true
        selectedDrawingButton.titleLabel?.lineBreakMode = NSLineBreakMode.byClipping
        selectedDrawingButton.layer.borderWidth = 1.0
        selectedDrawingButton.layer.borderColor = UIColor.gray.cgColor

        heightConstraint = constraints.first(where: {$0.firstAttribute == .height})
            ?? heightAnchor.constraint(equalToConstant: DrawingSwitcher.minHeight)
        heightConstraint?.isActive = true
        hideView()
    }

    func hideView() {
        DispatchQueue.main.async {
            self.contentView?.isHidden = true
            self.drawingsTableView?.isHidden = true
            self.updateFrame()
        }
    }

    func removeVenueMapDelegate() {
        if let venueMap = venueMap, let venueMapDelegate = venueMapDelegate {
            venueMap.removeVenueSelectionDelegate(venueMapDelegate)
            venueMap.removeDrawingSelectionDelegate(venueMapDelegate)
        }
    }

    func updateViewForSelectedVenue() {
        if venueMap == nil {
            hideView()
            return
        }

        if let venue = venueMap?.selectedVenue {
            updateDrawing(forVenue: venue)
        }
    }

    // Update this DrawingSwitcher with a new venue.
    func updateDrawing(forVenue venue: Venue) {
        // Make DrawingSwitcher invisible if there is less then 2 drawings in the venue.
        if venue.venueModel.drawings.count < 2 {
            hideView()
            return
        }

        DispatchQueue.main.async {
            // Set a name for the selected drawing to the title.
            self.contentView.isHidden = false
            let nameProp: Property? = venue.selectedDrawing.properties["name"]
            let name: String? = nameProp?.string
            self.selectedDrawingButton?.setTitle(name, for: UIControl.State.normal)

            // Get names of drawings and add them to the drawingNames variable.
            self.drawingNames.removeAll()
            let drawings: [VenueDrawing] = venue.venueModel.drawings
            for drawing in drawings {
                self.drawingNames.append(drawing.properties["name"]?.string ?? "")
            }

            self.drawingsTableView?.reloadData()
        }
    }

    public func setVenueMap(_ venueMap: VenueMap?) {
        // Remove old venue map delegates.
        removeVenueMapDelegate()
        // Set VenueMap for this DrawingSwitcher.
        if let venueMap = venueMap {
            self.venueMap = venueMap
            let delegate = DrawingSwitcherDelegate(drawingSwitcher: self)
            venueMap.addVenueSelectionDelegate(delegate)
            venueMap.addDrawingSelectionDelegate(delegate)
            venueMapDelegate = delegate
            updateViewForSelectedVenue()
        }
    }

    @IBAction private func onDrawingButtonClick(withSender: Any) {
        if drawingNames.count > 1 {
            // Show or hide the table view with drawings.
            drawingsTableView.isHidden = !drawingsTableView.isHidden
        } else {
            drawingsTableView.isHidden = true
        }

        updateFrame()
    }

    func updateFrame() {
        heightConstraint?.constant = drawingsTableView.isHidden
            ? DrawingSwitcher.minHeight
            : DrawingSwitcher.maxHeight
    }
}

extension DrawingSwitcher: UITableViewDataSource {
    public func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return drawingNames.count
    }

    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: UITableViewCell =
            tableView.dequeueReusableCell(withIdentifier: "DrawingSwitcherCellID", for: indexPath)

        cell.textLabel?.adjustsFontSizeToFitWidth = true
        cell.textLabel?.textAlignment = NSTextAlignment.center
        cell.textLabel?.text = drawingNames[indexPath.row]

        return cell
    }
}

extension DrawingSwitcher: UITableViewDelegate {
    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let drawingIndex: Int = indexPath.row
        if let venue = venueMap?.selectedVenue {
            // Set the selected drawing when a user clicks on the item in the table view.
            let drawing: VenueDrawing = venue.venueModel.drawings[drawingIndex]
            venue.selectedDrawing = drawing
            // Hide the table view.
            drawingsTableView.isHidden = true
            updateFrame()
        }
    }
}

File: examples/latest/navigate/android/IndoorMap/app/src/main/java/com/here/sdk/examples/venues/DrawingSwitcher.java
/*
 * Copyright (C) 2019-2024 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.sdk.examples.venues;

import android.app.Activity;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.GradientDrawable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.ListAdapter;
import android.widget.ListView;

import com.here.sdk.venue.control.VenueDrawingSelectionListener;
import com.here.sdk.venue.control.VenueSelectionListener;
import com.here.sdk.venue.control.Venue;
import com.here.sdk.venue.control.VenueMap;
import com.here.sdk.venue.data.VenueDrawing;
import com.here.sdk.venue.data.Property;
import com.here.sdk.venue.data.VenueModel;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

// Allows to select a drawing inside a venue trough UI.
public class DrawingSwitcher implements View.OnClickListener, AdapterView.OnItemClickListener {
    private Context context;
    private VenueMap venueMap;
    private Venue venue;
    private ImageButton titleView;
    private ListView listView;
    private boolean collapsed;

    public DrawingSwitcher(Context context, ImageButton imageButton, ListView listView) {

        this.context = context;
        this.titleView = imageButton;
        this.listView = listView;
        titleView.setVisibility(View.GONE);
        listView.setVisibility(View.GONE);
        titleView.setOnClickListener(this);
        listView.setOnItemClickListener(this);

        collapsed = true;
    }

    private final VenueSelectionListener venueSelectionListener =
            (deselectedController, selectedController) -> {
        // Update DrawingSwitcher with a new selected venue.
        DrawingSwitcher.this.setVenue(selectedController);
        if (selectedController != null) {
            // Update DrawingSwitcher with a new selected drawing.
            DrawingSwitcher.this.onDrawingSelected(selectedController.getSelectedDrawing());
        }
    };

    private final VenueDrawingSelectionListener drawingSelectionListener =
            (venue,
             deselectedController,
             selectedController) ->
                    // Update DrawingSwitcher with a new selected drawing.
                    DrawingSwitcher.this.onDrawingSelected(selectedController);

    @Override
    protected void finalize() throws Throwable {
        removeListeners();
        super.finalize();
    }

    private void removeListeners() {
        if (venueMap != null) {
            venueMap.remove(venueSelectionListener);
            venueMap.remove(drawingSelectionListener);
        }
    }

    public void setVenueMap(VenueMap map) {
        // Remove old venue map listeners.
        removeListeners();
        // Set VenueMap for this DrawingSwitcher.
        venueMap = map;
        venueMap.add(venueSelectionListener);
        venueMap.add(drawingSelectionListener);
    }

    @Override
    public void onClick(View v) {
        // Open of hide the list with drawings when a user clicks on the title.
        setCollapsed(!collapsed);
    }

    @Override
    public void onItemClick(AdapterView<?> parent, final View view, int position, long id) {
        VenueModel venueModel = venue.getVenueModel();
        // Set the selected drawing when a user clicks on the item in the list.
        venue.setSelectedDrawing(venueModel.getDrawings().get(position));
        // Hide the list.
        setCollapsed(true);
    }

    // Gets name of the drawing.
    private String getDrawingName(VenueDrawing drawing) {
        String name = "";
        Map<String, Property> properties = drawing.getProperties();
        Property property = properties.get("name");
        if (property != null) {
            name = property.getString();
        }
        return name;
    }

    // Sets a new venue for this DrawingSwitcher.
    private void setVenue(Venue venue) {
        this.venue = venue;
        int count = 0;
        int height = 0;
        ListAdapter listAdapter;
        if (this.venue != null) {
            // Get names of drawings.
            VenueModel venueModel = this.venue.getVenueModel();
            List<VenueDrawing> drawings = venueModel.getDrawings();
            count = drawings.size();
            final List<String> drawingNames = new ArrayList<>();
            for (VenueDrawing drawing : drawings) {
                drawingNames.add(getDrawingName(drawing));
            }

            // Set a name of the selected drawings
            String selectedDrawingName = getDrawingName(this.venue.getSelectedDrawing());

            // Set a new adapter with the new list of drawing's names.
            final StringArrayAdapter adapter =
                    new StringArrayAdapter(context, drawingNames);
            ViewGroup.MarginLayoutParams margin = (ViewGroup.MarginLayoutParams) listView.getLayoutParams();
            listView.setAdapter(adapter);
            listAdapter = listView.getAdapter();
            for(int i=0; i<count; i++){
                View childView = listAdapter.getView(i, null, listView);
                childView.measure(View.MeasureSpec.makeMeasureSpec(0,
                        View.MeasureSpec.UNSPECIFIED), View.MeasureSpec
                        .makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED));
                height += childView.getMeasuredHeight();
            }
            Log.d("DrawingSwitcher", "marginTop:" + margin.topMargin + "height: " + height);
            margin.setMargins(margin.leftMargin, 1500 - height, margin.rightMargin, margin.bottomMargin);
            setCollapsed(true);
            setVisible(true);
        }

    }

    // Updates the title with a new selected drawing.
    private void onDrawingSelected(final VenueDrawing selectedDrawing) {
        if (selectedDrawing == null) {
            return;
        }
        String selectedDrawingName = getDrawingName(selectedDrawing);
    }

    public void setVisible(boolean visible) {
        titleView.setVisibility(visible ? View.VISIBLE : View.GONE);
        if (!collapsed) {
            listView.setVisibility(visible ? View.VISIBLE : View.GONE);
        }
    }

    private void setCollapsed(boolean collapsed) {
        if (collapsed) {
            listView.setVisibility(View.GONE);
        } else {
            listView.setVisibility(View.VISIBLE);
        }
        this.collapsed = collapsed;
    }
}
